groupBy shall emit SubFlow

rkuhn opened the pull request:
fixes #19005 

will squash/rebase after review

drewhk commented on akka-docs-dev/rst/java/stream-cookbook.rst:
```diff
@@ -107,23 +107,30 @@ Implementing reduce-by-key
-
-.. includecode:: ../../../akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/cookbook/RecipeReduceByKey.java#word-count
+To count the words, we need to process the stream of streams (the actual groups
+containing identical words). ``GroupBy`` returns a :class:`SubSource`, which
```
weird capitalization of GroupBy since in code it is small-caps. I usually try to refactor these kind of sentences so that the first word is not code to sidestep this issue.

rkuhn replied:
fixed

drewhk commented on akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeReduceByKey.scala:
```diff
@@ -17,22 +17,15 @@ class RecipeReduceByKey extends RecipeSpec {
+        // split the words into separate streams first
+        .groupBy(identity)
+        // add counting logic to the streams
+        .fold(("", 0)) {
```
another reason why a non-terminal `fold` is useful. Thanks @viktorklang for doing the necessary refactor!

viktorklang replied:
The Klang abides :)

drewhk commented on akka-stream-tck/src/test/scala/akka/stream/tck/GroupByTest.scala:
```diff
@@ -17,7 +17,12 @@ class GroupByTest extends AkkaPublisherVerification[Int] {
-        Source(iterable(elements)).groupBy(elem ⇒ "all").map { case (_, group) ⇒ group }.runWith(Sink.head)
+        Source(iterable(elements))
+          .groupBy(elem ⇒ "all")
+          .prefixAndTail(0)
```
shouldn't this be called `lift` or something? Also, this method should take some kind of licence ;)

rkuhn replied:
That is precisely the reason for:

* not documenting this purpose
* making it involve some boilerplate
* and in general being non-obvious

:-)

drewhk commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGroupBySpec.scala:
```diff
@@ -13,8 +13,21 @@ import akka.stream.testkit.Utils._
+  import language.higherKinds
+
+  implicit class Lift[M](val f: SubFlow[Int, M, Source[Int, M]#Repr, RunnableGraph[M]]) extends AnyVal {
+    def lift(key: Int ⇒ Int) = f.prefixAndTail(1).map(p ⇒ key(p._1.head) -> (Source.single(p._1.head) ++ p._2)).mergeBack(1)
```
here is my lift ;)

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/SubFlowImpl.scala:
```diff
@@ -0,0 +1,35 @@
+import language.higherKinds
+
+object SubFlowImpl {
+  trait MergeBack[In, F[+_]] {
```
`F[+_]` lol wat?

rkuhn replied:
It’s just a monoid in the category of endofunctors … (no, it is actually a covariant type constructor).

viktorklang replied:
It's a type constructor named F with a single type parameter which is covariant.

:sunglasses:

drewhk replied:
got it.

drewhk commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -733,12 +732,20 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
+   *
+   * It is important to note that the substreams also propagate back-pressure as
+   * any other stream, which means that blocking one substream will block the `groupBy`
+   * operator itself&mdash;and thereby all substreams&mdash; once all internal or
```
disagree with the `&mdash`. While it might improve scaladocs, it is not that uncommon to read these comments in code, where it is annoying.

rkuhn replied:
What do you propose? (and there is an extraneous space character on this line as well … )

drewhk replied:
can't you just paste the proper UTF-8 character? Or does Javadoc trip on that?

rkuhn replied:
Using the UTF-8 character “works for me” (tm) on Chrome, and I’ll probably just do that, but old browsers will choke on this (and display a replacement glyph or something weird).

drewhk replied:
let's just say "ancient, no longer supported browsers" instead of "old browsers" ;) I don't think we need to worry about that. If someone reports this problem we will fix it, but I highly doubt that this will be a problem nowadays.

drewhk commented on akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala:
```diff
@@ -0,0 +1,865 @@
+   * WARNING: currently only a breadth of 1 is actually implemented! Beware of deadlocks!
+   */
+  // FIXME remove doc warning once fixed
+  def mergeBack(breadth: Int): Flow[In, Out, Mat] =
```
shouldn't we create a `concat()` method? We tried to be rather restrictive in the past and making certain dangerous methods/combinations harder to access but then later we reverted many of them. `concat()` conveys intent more than `mergeBack(1)`

rkuhn replied:
Okay, so here’s what I’ll do:

* remove the parameter and rename to `mergeSubstreams`
* add `concatSubstreams` method
* add `maximumSubstreams` parameter to `groupBy`

drewhk replied:
`concatSubstreams` should take a `String` parameter that must have the exact value of "I know what I am doing." ;)

viktorklang replied:
It's not `maximumSubstreams` though. It's `maxConcurrentSubstreams`

On Wed, Nov 25, 2015 at 12:24 PM, Roland Kuhn <notifications@github.com>
wrote:

> In akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala
> <https://github.com/akka/akka/pull/19006#discussion_r45855773>:
>
> > +class SubFlow[-In, +Out, +Mat](delegate: scaladsl.SubFlow[Out, Mat, scaladsl.Flow[In, Out, Mat]#Repr, scaladsl.Sink[In, Mat]]) {
> > +
> > +  /** Converts this Flow to its Scala DSL counterpart */
> > +  def asScala: scaladsl.SubFlow[Out, Mat, scaladsl.Flow[In, Out, Mat]#Repr, scaladsl.Sink[In, Mat]] @uncheckedVariance = delegate
> > +
> > +  /**
> > +   * Flatten the sub-flows back into the super-flow, performing a merge of the
> > +   * given breadth. For concatenation of the flows use a breadth of 1 but beware
> > +   * that this will lead to deadlocks if the sub-flows need to make progress
> > +   * concurrently (as is generally the case when using `groupBy` on an unordered
> > +   * input stream).
> > +   *
> > +   * WARNING: currently only a breadth of 1 is actually implemented! Beware of deadlocks!
> > +   */
> > +  // FIXME remove doc warning once fixed
> > +  def mergeBack(breadth: Int): Flow[In, Out, Mat] =
>
> Okay, so here’s what I’ll do:
>
>    - remove the parameter and rename to mergeSubstreams
>    - add concatSubstreams method
>    - add maximumSubstreams parameter to groupBy
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19006/files#r45855773>.
>



-- 
Cheers,
√

drewhk replied:
Since groupBy is not able to close any group before the primary stream is closed, it is actually `maximumSubstreams`.

viktorklang replied:
Indeed! I stand corrected.

viktorklang commented on akka-docs-dev/rst/java/migration-guide-1.0-2.x-java.rst:
```diff
@@ -422,7 +422,7 @@ should be replaced by:
 
 SynchronousFileSource and SynchronousFileSink
-============================================
+=============================================
```
GAME CHANGER

drewhk replied:
Well, now it will output proper HTML ;) (yes, RST actually does care of the length of those bars)

ktoso replied:
Oh I didn't realise it actually breaks when not aligned. I always fix them because of my OCD anyway :wink:

drewhk replied:
you don't use the RST plugin in IDEA? It shows misaligned header lines.

ktoso replied:
I? Sure I do; so I always fix those. But never realised it's a real rendering problem that I'm fixing ;)

drewhk commented on akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala:
```diff
@@ -0,0 +1,865 @@
+  def log(name: String): SubFlow[In, Out, Mat] =
+    this.log(name, ConstantFun.javaIdentityFunction[Out], null)
+
+}
```
eh, +865 lines of duplication... Would be nice at least to have some boilerplate plugin like magic to generat these...

ktoso replied:
yeah, this is becoming really crazy... :( 
Make it a ticket, if I'm a bit more free I can try out something.

ktoso replied:
https://github.com/akka/akka/issues/19019

drewhk commented on akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala:
```diff
@@ -0,0 +1,865 @@
+  def log(name: String): SubSource[Out, Mat] =
+    this.log(name, ConstantFun.javaIdentityFunction[Out], null)
+
+}
```
ah, 1700 more lines, aaargh :(

2beaucoup commented:
There are a lot of `prefixAndTail(1)`s here. Maybe introduce `headAndTail` which returns `(Out, Source[U, Unit])` instead of `(immutable.Seq[Out], Source[U, Unit])` and which spares users the hassle of sequence matching? ;)

drewhk commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -915,8 +913,36 @@ trait FlowOps[+Out, +Mat] {
+        deprecatedAndThen[Source[Out, Unit]](GroupBy(f.asInstanceOf[Any ⇒ Any]))
+          .map(_.via(flow).via(makeEager) /* needed to start generating some demand */ )
+          .map(s ⇒ Source(s.runWith(Sink.publisher(false)))) // need to be eager as a workaround for now
+          .buffer(4, OverflowStrategy.Backpressure) // needed to make reduceByKey recipe work
```
? That is weird.

drewhk replied:
Ah, to feed enough makeEager into the map above?

rkuhn replied:
Nope, it is an artifact of using the totally lazy ConcatAll for now: it will not pull any more elements (nested streams) out of the prematerialization `map()` than what its input buffer demands (4) while the sample has 5 substreams … We must fix this for real before doing a release, which means actually implementing a flatMapMerge with configurable breadth.

drewhk replied:
that's what I said above. This buffer makes sure that all subgroups pass the map above, that attaches the makeEager stage.

viktorklang replied:
I am concerned by `breadth` as it becomes "dynamically typed" and depending on the Flows before it.
What other options are there?

Case in point:

groupBy(…).groupBy(…).mergeBack(2)

and then someone adds or removes a groupBy but forgets to change the magic number.

rkuhn replied:
@viktorklang Nothing about the structure is dynamic here, the breadth parameter only determines how many substreams are merged concurrently. Your “case in point” would not compile in a context where a normal stream is expected because every `groupBy` needs to be matched by exactly one `mergeBack`.

viktorklang replied:
Ah, ok, then if it is how many streams are merged in concurrently I think it would be nice to have a cohesive name for that parameter value, as it is related to mapAsync/mapAsyncUnordered etc. so mergeBack(1) == concat

clockfly replied:
+1 on Viktor.
mergeBack(1) don't looks good, it may create confusion with merge.
Maybe using name flatten or concat is better.

rkuhn replied:
If you see `mergeBack(1)` in code you are almost always looking at a potential deadlock, so I disagree with the notion that there may be confusion here—what we do is to make the reasonable thing simple and discoverable.

OTOH I am really having doubts on this limit: instead of having people choose a limit beyond which the stream just deadlocks, why don’t we keep the mergeBack unlimited and have people think about limits when invoking groupBy?

viktorklang replied:
Having an Int-param on groupBy to denominate max number of concurrent groups sounds like a stroke of genius.

rkuhn replied:
Let’s just hope that the genius doesn’t die of a stroke soon … ;-)

drewhk replied:
Not to take away the *genius* marker from @rkuhn but we actually discussed this in the past, it just never have been applied (if I recall because we wanted to fix stream-of-streams stuff properly, once and for all, so we postponed it).

drewhk commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -915,8 +913,36 @@ trait FlowOps[+Out, +Mat] {
+    val merge = new SubFlowImpl.MergeBack[Out, Repr] {
+      override def apply[T](flow: Flow[Out, T, Unit], breadth: Int): Repr[T] =
+        deprecatedAndThen[Source[Out, Unit]](GroupBy(f.asInstanceOf[Any ⇒ Any]))
+          .map(_.via(flow).via(makeEager) /* needed to start generating some demand */ )
```
because of the mapConcat right? Otherwise it will not start consuming all subgroups?

drewhk commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -915,8 +913,36 @@ trait FlowOps[+Out, +Mat] {
+          .flatMapConcat(identity)
+          .asInstanceOf[Repr[T]]
+      /*
+       * FIXME remove all those commented workarounds above by implementing flatMapMerge(breadth)
```
I see.

viktorklang commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -915,8 +913,36 @@ trait FlowOps[+Out, +Mat] {
+          .map(_.via(flow).via(makeEager) /* needed to start generating some demand */ )
+          .map(s ⇒ Source(s.runWith(Sink.publisher(false)))) // need to be eager as a workaround for now
+          .buffer(4, OverflowStrategy.Backpressure) // needed to make reduceByKey recipe work
+          .flatMapConcat(identity)
```
Use proper identity, the one of constant fun.

rkuhn replied:
This code is broken in more important ways and will be replaced in the next PR :-)

viktorklang commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -959,8 +997,20 @@ trait FlowOps[+Out, +Mat] {
+      override def apply[T](flow: Flow[Out, T, Unit], breadth: Int): Repr[T] =
+        deprecatedAndThen[Source[Out, Unit]](Split.when(p.asInstanceOf[Any ⇒ Boolean]))
+          .map(_.via(flow))
+          .flatMapConcat(identity)
```
Find and replace use of the wrong identity :)

drewhk commented:
Looks very nice so far!

rkuhn committed:
!str #19005 make groupBy et al return a SubFlow

A SubFlow (or SubSource) is not a Graph, it is an unfinished builder
that accepts transformations. This allows us to capture the substreams’
transformations before materializing the flow, which will be very
helpful in fully fusing all operators.

Another change is that groupBy now requires a maxSubstreams parameter in
order to bound its resource usage. In exchange the matching merge can be
unbounded. This trades silent deadlock for explicit stream failure.

This commit also changes all uses of Predef.identity to use `conforms`
and removes the HTTP impl.util.identityFunc.

rkuhn commented:
documented, squashed and rebased

ktoso commented on akka-docs-dev/rst/java/stream-cookbook.rst:
```diff
@@ -107,23 +107,30 @@ Implementing reduce-by-key
+it cannot consume the resulting elements within its resource bounds. The
+consequence is that the stream will deadlock.
+
+.. includecode:: ../../../akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/cookbook/RecipeReduceByKeyTest.java#word-count
```
tempted to add a 1 sentence summary in a warning block here?

ktoso commented on akka-docs-dev/rst/java/stream-cookbook.rst:
```diff
@@ -133,12 +140,13 @@ By extracting the parts specific to *wordcount* into
-  Please note that the reduce-by-key version we discussed above is sequential, in other words it is **NOT** a
+  Please note that the reduce-by-key version we discussed above is sequential
+  in reading the overall input stream, in other words it is **NOT** a
   parallelization pattern like mapReduce and similar frameworks.
```
MapReduce or map-reduce?

ktoso commented on akka-docs-dev/rst/java/stream-cookbook.rst:
```diff
@@ -150,12 +158,12 @@ Sometimes we want to map elements into multiple groups simultaneously.
 
 * first, using a function ``topicMapper`` that gives a list of topics (groups) a message belongs to, we transform our
-  stream of ``Message`` to a stream of ``(Message, Topic)`` where for each topic the message belongs to a separate pair
+  stream of ``Message`` to a stream of :class:`Pair<Message, Topic>`` where for each topic the message belongs to a separate pair
```
good catch, thanks

ktoso commented on akka-docs-dev/rst/scala/code/docs/MigrationsScala.scala:
```diff
@@ -116,9 +116,24 @@ class MigrationsScala extends AkkaSpec {
+        Flow[Int]
+          .groupBy(2, _ % 2) // the first parameter sets max number of substreams
+          .map(_ + 3)
+          .concatSubstreams
```
I'm late to the naming party but wanted to say it's a great name :+1:

ktoso commented on akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/cookbook/RecipeReduceByKeyTest.java:
```diff
@@ -46,27 +53,24 @@ public void work() throws Exception {
+            // split the words into separate streams first
+          .groupBy(MAXIMUM_DISTINCT_WORDS, i -> i)
+          // add counting logic to the streams
+          .fold(new Pair<>("", 0), (pair, elem) -> new Pair<>(elem, pair.second() + 1))
```
In theory we have `Pair.create` so you don't have to do the diamond.

rkuhn replied:
But `Pair.create` does not look nicer or save characters, so I prefer the obvious notation.

2beaucoup replied:
If it only would have been named `Pair.of`... :(

hepin1989 replied:
new api in jdk 9 will make use of the `of`
http://openjdk.java.net/jeps/269

ktoso replied:
I am a big fan of `of`. I think we should add it (once streams are on master, that'll be much simpler)

hepin1989 replied:
And more `of` here https://github.com/jOOQ/jOOL

ktoso commented on akka-stream-testkit/src/main/scala/akka/stream/testkit/StreamTestKit.scala:
```diff
@@ -42,7 +42,7 @@ object TestPublisher {
 
   /**
-   * Publisher that signals error to subscribers immediately, before handing out subscription.
+   * Publisher that signals error to subscribers immediately after handing out subscription.
```
good catch

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -757,9 +764,11 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
    * '''Cancels when''' downstream cancels and all substreams cancel
    *
+   * @param maxSubstreams configures the maximum number of substreams (keys)
+   *        that are supported; if more distinct keys are encountere then the stream fails
```
s/encountere/encountered

viktorklang commented on akka-docs-dev/rst/java/stream-cookbook.rst:
```diff
@@ -107,23 +107,30 @@ Implementing reduce-by-key
+Therefore only a finite number of substreams can be active at any given time.
+If the ``groupBy`` operator emits more than this number then the stream cannot
+continue: in order to write the new word to its new substream there needs to be
+demand on that substream and ``mergeBack`` will not create this demand because
```
Strongly against `mergeBack` as a name, especially if the method does not exist ;)

viktorklang commented on akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeReduceByKey.scala:
```diff
@@ -54,15 +47,17 @@ class RecipeReduceByKey extends RecipeSpec {
+            case (Some((key, out)), elem) =>
+              Some((key, fold(out, elem)))
+          }
+          .map(_.get)
```
?

rkuhn replied:
WDYM?

viktorklang replied:
Option.get is always a smell, if you want it to fail on None use getOrElse and throw a more meaningful exception?

drewhk replied:
In this case it emulates `reduce` which throws on an empty collection. Here it throws on an empty stream.

rkuhn replied:
What you guys are overlooking is that this cannot ever be empty, by construction. `reduce` is more appropriate but does not exist yet.

viktorklang replied:
Yes, if you re-read what I wrote, you'll notice that I have nothing against throwing exceptions but what I was against was A) the use of Option.get and B) the generated exception from using Option.get

Option.get should not be encouraged, and in this case the error message ought to be different than: java.util.NoSuchElementException: None.get

drewhk replied:
I know it cannot be empty, I wrote this sample :) I just pointed out that it does exactly what reduce: works only with non-empty collections. I did not imply that the substream can be empty.

rkuhn replied:
@viktorklang we’re arguing over the empty set, especially in view of #19020.

viktorklang replied:
@rkuhn :+1:

viktorklang commented on akka-docs-dev/rst/scala/stream-testkit.rst:
```diff
@@ -4,60 +4,89 @@
 
-The same strategy can be applied for sources as well. In the next example we have a source that produces an infinite stream of elements. Such source can be tested by asserting that first arbitrary number of elements hold some condition. Here the :code:`grouped` combinator and :code:`Sink.head` are very useful.
+The same strategy can be applied for sources as well. In the next example we
+have a source that produces an infinite stream of elements. Such source can be
```
Should we consistently use `unbounded` iso `infinite`?

rkuhn replied:
good point, please track as a separate ticket

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -533,6 +533,14 @@ final class Source[+Out, +Mat](delegate: scaladsl.Source[Out, Mat]) extends Grap
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
```
so... about that genjavadoc improvement... :stuck_out_tongue_winking_eye:

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -541,16 +549,38 @@ final class Source[+Out, +Mat](delegate: scaladsl.Source[Out, Mat]) extends Grap
 
   /**
-   * Transform each input element into a sequence of output elements that is
+   * Transform each input element into an `Iterable` of output elements that is
```
technically into a `java.util.List[T]]`

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -541,16 +549,38 @@ final class Source[+Out, +Mat](delegate: scaladsl.Source[Out, Mat]) extends Grap
+   * being used as an output sequence. Otherwise the stream may fail with
+   * `ConcurrentModificationException` or other more subtle errors may occur.
+   *
+   * The returned `Iterable` MUST NOT contain `null` values,
```
java.util.List[T]], or change the signature to be Iterable?

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala:
```diff
@@ -0,0 +1,868 @@
+   * }}}
+   *
+   * The materialized value of the combined [[Flow]] will be the materialized
+   * value of the current flow (ignoring the other Flow’s value), use
```
other => inner?

rkuhn replied:
no, the current is inner as well

viktorklang commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowConcatAllSpec.scala:
```diff
@@ -44,7 +44,13 @@ class FlowConcatAllSpec extends AkkaSpec {
+        .prefixAndTail(0)
+        .map(_._2)
+        .concatSubstreams
+        .flatMapConcat(ConstantFun.scalaIdentityFunction)
```
I find this code confusing. :( (i.e. the double concats)

rkuhn replied:
Hmm, I don’t see the problem: there are two layers of substreaming, so there must be two layers of flattening as well. What is confusing?

viktorklang replied:
Since `concatSubstreams` and `flatMapConcat(identity)` are isomorphic, it looks weird to have both, and to have them adjacent, don't you agree?

drewhk replied:
But what is your suggestion? They are different operations with different safety and operational requirements.

viktorklang replied:
Looking at the code above, how do I as the user or reader of the code by any means elicit these safety and operational requirements?

drewhk replied:
You ask questions, but don't provide an answer. You kind of suggest that there is, but you don't give it. If you have an answer, then please state it as a constructive recommendation, not an open ended question.

rkuhn replied:
These operations are not isomorphic, they even have different types. And users will not—they shall not—use prefixAndTail unless they know what they are doing.

viktorklang replied:
No, that was an honest question. As the designers of this API, we need to be able to answer it, and right now I can't if someone were to ask me. So I'm asking the author.

drewhk replied:
```
.prefixAndTail(0)
.map(_._2)
.concatSubstreams
```
is a funky way to way to lift the substreams to actual elements of the stream. It is funky because it is dangerous therefore we don't provide sugar for it. The additional `flatMapConcat` is part of the integration test which is to see if splitWhen and concatAll works together.

viktorklang replied:
@drewhk Thanks! Alright, so if it is code we don't want people to write anyway then it doesn't really count towards it that it looks confusing. Thanks for educating me!

rkuhn replied:
Right, the only reason for this code to exist is that I changed the API without changing what the test does, but the test does no longer use the API in a realistic or idiomatic way. Nevertheless it does test all those important properties.

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1245,7 +1294,26 @@ trait FlowOps[+Out, +Mat] {
+   * }}}
+   * The materialized value of the combined [[Sink]] will be the materialized
+   * value of the current flow (ignoring the given Sink’s value), use
+   * [[Flow#toMat[Mat2* toMat]] if a different strategy is needed.
```
whoa, this ref works?

rkuhn replied:
no, of course not (yes, it does work for ScalaDoc, but not for JavaDoc)

rkuhn replied:
which is totally okay since they are separate, but thanks for the poking: the Java version had not yet been outfitted with the drawing!

ktoso replied:
great, the drawings are very helpful :+1:

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1259,29 +1327,122 @@ trait FlowOps[+Out, +Mat] {
+  def alsoToMat[Mat2, Mat3](that: Graph[SinkShape[Out], Mat2])(matF: (Mat, Mat2) ⇒ Mat3): ReprMat[Out, Mat3] =
+    viaMat(alsoToGraph(that))(matF)
+
+  private[akka] def transformMaterializing[T, M](mkStageAndMaterialized: () ⇒ (Stage[Out, T], M)): ReprMat[T, M] =
```
INTERNAL API?

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/SubFlow.scala:
```diff
@@ -0,0 +1,40 @@
+ * SubFlows cannot contribute to the super-flow’s materialized value since they
+ * are materialized later, during the runtime of the flow graph processing.
+ */
+trait SubFlow[+Out, +Mat, +F[+_], C] extends FlowOps[Out, Mat] {
```
(<3 scala's code reusability)

2beaucoup commented:
> There are a lot of prefixAndTail(1)s here. Maybe introduce headAndTail which returns (Out, Source[U, Unit]) instead of (immutable.Seq[Out], Source[U, Unit]) and which spares users the hassle of sequence matching? ;)

WDYT @rkuhn?

rkuhn commented:
`prefixAndTail` is the single most dangerous combinator in the whole library that remains after these changes, so making it any more attractive is not a good idea: if people really need to formulate dynamic stream processing topologies then they’ll need to go through some undocumented boilerplate rituals. We do want to make the alternatives as easily accessible as possible, though, which this PR is one part of.

ktoso commented:
Needs a rebase and validation failed on compiling java tests:

```
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-stream-tests/src/test/java/akka/stream/javadsl/SourceTest.java:67: error: method mapConcat in class Source<Out,Mat> cannot be applied to given types;

```

rkuhn commented:
Yes, I stashed that and could potentially finish it now that FlattenMerge is there, but I’m fusing all the things right now :-)

ktoso commented:
Roger that, thanks for the update (going through all PRs today)
